plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java

```diff
@@ -52,59 +51,79 @@
   private static final Logger log = LoggerFactory.getLogger(PyPiIndexUtils.class);
 
   /**
-   * Returns a map (in original order of appearance) of the files and associated paths extracted from the index.
+   * From PEP_053:
+   * A repository MAY include a data-requires-python attribute on a file link.
+   * This exposes the Requires-Python metadata field, specified in PEP 345, for the corresponding release.
+   * Where this is present, installer tools SHOULD ignore the download when installing
+   * to a Python version that doesn't satisfy the requirement.
    */
-  static Map<String, String> extractLinksFromIndex(final InputStream in) throws IOException {
+  private static final String PYPI_REQUIRES_PYTHON = "data-requires-python";
+
+  /**
+   * Return a list (in original order of appearance) of PyPiLinkInfo extracted from the index.
+   */
+  static List<PyPiLink> extractLinksFromIndex(final InputStream in) throws IOException {
     checkNotNull(in);
-    Map<String, String> results = new LinkedHashMap<>();
+    List<PyPiLink> results = new LinkedList<>();
     try (Reader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
       String html = CharStreams.toString(reader);
       Document document = Jsoup.parse(html);
       Elements links = document.select("a[href]");
       for (Element link : links) {
         String file = link.text().trim();
         String path = link.attr("href");
-        if (!results.containsKey(file)) {
-          results.put(file, path);
-        }
+        String requiresPython = link.attr(PYPI_REQUIRES_PYTHON);
+        results.add(new PyPiLink(file, path, requiresPython));
       }
     }
     return results;
   }
 
+
+  /**
+   * Convert a link to a map ready to be used for "pypi-index.vm".
+   */
+  static ImmutableMap<String, String> indexLinkToMap(PyPiLink link) {
+    return ImmutableMap.of(
+            "link", makePackageLinkRelative(link.getLink()),
+            "file", link.getFile(),
+            "data-requires-python", link.getDataPythonRequires()
+    );
+  }
+
+  /**
+   * Convert a link to a map ready to be used for "pypi-root-index.vm".
+   */
+  static ImmutableMap<String, String> rootIndexLinkToMap(PyPiLink link) {
```

Comment:
According to Sonatype code style method params should be final. would u please set PyPiLink link as final?